File: Data Structures & Algorithms/valid-tree/submission-3.cpp
class Solution {
private:
    void dfs(int node, int &count, const vector<vector<int>>& adj, vector<bool>& visited){
        visited[node] = true;
        count++;
        for (int n: adj[node]){
            if (!visited[n]){
                dfs(n, count, adj, visited);
            }
        }
        return;
    }    
public:
    bool validTree(int n, vector<vector<int>>& edges) {
        int length = edges.size();
        int count = 0;
        if (length != n - 1){
            return false;
        }

        vector<bool> visited(n, false);
        vector<vector<int>> adj(n);
        for (const auto& edge : edges) {
        int a = edge[0];
        int b = edge[1];
        adj[b].push_back(a); 
        adj[a].push_back(b);
        }
        dfs(0, count, adj, visited);
        if (count == n){
            return true;
        }
        return false;
        
    }
};

File: Data Structures & Algorithms/redundant-connection/submission-3.py
class Solution:
    def findRedundantConnection(self, edges: list[list[int]]) -> list[int]:
        n = len(edges)
        adj = [[] for _ in range(n + 1)]

        def dfs(curr, target, visited):
            if curr == target:
                return True
            if curr in visited:
                return False    
            visited.add(curr)    

            for c in adj[curr]:
                if c not in visited:
                    if dfs(c, target, visited):
                        return True 
            return False                   

        for a, b in edges:
            visited = set()
            if dfs(a, b, visited):
                return [a, b]
            adj[a].append(b)
            adj[b].append(a)                